src/glb.rs

```diff
@@ -55,71 +126,134 @@ impl<'a> Glb<'a> {
+                                    buf: &'a mut Vec<u8>) -> Result<Self, Error> {
+        let header = Header::from_reader(&mut reader).map_err(Error::Glb)?;
+        match header.version {
+            2 => {
+                let glb_len = header.length - Header::size_of() as u32;
+                buf.clear();
+                buf.reserve(glb_len as usize);
+                // SAFETY: We are doing unsafe operation on a user-supplied
+                // container!  Make sure not to expose user to uninitialized
+                // data if an error happens during reading.
+                //
+                // It is guaranteed by reserve's implementation that the reserve
+                // call will make buf's capacity _at least_ buf_len.
+                //
+                // We do not read contents of the Vec unless it is fully
+                // initialized.
+                unsafe { buf.set_len(glb_len as usize) };
+                if let Err(e) = reader.read_exact(buf).map_err(GlbError::IoError) {
+                    // SAFETY: It is safe to not run destructors because u8 has
+                    // none.
+                    unsafe { buf.set_len(0) };
+                    Err(Error::Glb(e))
+                } else {
+                    Self::from_v2(buf)
+                       .map(|(json, bin)| Glb { header, json, bin })
+                       .map_err(Error::Glb)
+                }
             }
-            *x
-        };
+            x => Err(Error::Glb(GlbError::Version(x)))
+        }
+    }
 
-        offset += size_of::<Header>() as isize;
-        let (json_chunk, json_chunk_length) = unsafe {
-            let x: *const ChunkInfo = transmute(ptr.offset(offset));
-            offset += size_of::<ChunkInfo>() as isize;
-            if (*x).length as usize > (data.len() as isize - offset) as usize {
-                return err("JSON chunkLength exceeds length of data");
-            }
-            if &(*x).kind != b"JSON" {
-                return err("JSON chunkType is not `JSON`");
-            }
-            let begin = offset as usize;
-            let end = begin + (*x).length as usize;
-            (begin..end, (*x).length as usize)
-        };
+    fn from_v2(mut data: &'a [u8]) -> Result<(&'a [u8], Option<&'a [u8]>), GlbError> {
+        let (json, mut data) = ChunkHeader::from_reader(&mut data)
+            .and_then(|json_h| if let ChunkType::Json = json_h.ty {
+                Ok(json_h)
+            } else {
+                Err(GlbError::ChunkType(json_h.ty))
+            })
+            .and_then(|json_h| if json_h.length as usize <= data.len() {
+                Ok(json_h)
+            } else {
+                Err(GlbError::ChunkLength {
+                    ty: json_h.ty,
+                    length: json_h.length,
+                    length_read: data.len(),
+                })
+            })
+            // PANIC: We have verified that json_h.length is no greater than
```

Comment:
I'm a bit worried to be reading `PANIC` in big letters in the source code. Does this actually `panic!`?

Reply:
No, it explains why it doesn't.

Reply:
```rust
/// FIRE MISSILES: This function does not fire any missiles.
fn foo() {}
```